### ISSUE TITLE
Array transmutes (akin to reinterpret_cast)

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#21998]
Say I have the following:

```chpl
var arr64 : [0..<2] uint(64) = [0x1122334455667788, 0xAABBCCDDEEFF0011];
var arr32 : [0..<4] uint(32);
```

And I want to do a bit-level copy from `arr64` to `arr32`.

I could do this in an elementwise fashion like this:

```chpl
var arr64 : [0..<2] uint(64) = [0x1122334455667788, 0xAABBCCDDEEFF0011];
var arr32 : [0..<4] uint(32);

for i in 0..<arr64.size {
  var low  =  (arr64[i] & 0x00000000FFFFFFFF)        : uint(32);
  var high = ((arr64[i] & 0xFFFFFFFF00000000) >> 32) : uint(32);
  arr32[i*2] = low; arr32[i*2+1] = high;
}

for val in arr32 do writef("%xu ", val);
writeln();
```

But I'd rather just be able to do it as a oneliner type of operation (something akin to using a `reinterpret_cast` in C++).

I'm not sure what syntax would be best for this, but maybe maybe something like this:

`arr32 = transmuteArray(arr64, uint(32));`

To further complicate this issue. The use case I'm really interested in is when one of the arrays is on a GPU, in other words something like this (assume I'm in array-on-device mode):

```chpl
var arr64 : [0..<2] uint(64) = [0x1122334455667788, 0xAABBCCDDEEFF0011];
on here.gpus[0] {
  var arr32 : [0..<4] uint(32);
  arr32 = arry4;
}
```

The CUDA version of this would turn into a single call to `cudaMemcpy` where the type information gets erased since everything is casted to `void*`.

In this case writing out the for loop I show above that does an elementwise copy is even more terrible as it will introduces a communication call for every element of the 64-bit array. I imagine if we got this working for the "CPU array to CPU array" version the GPU version would fall out but as part of this issue it would be nice to double check that this is the case and if not fork a new issue for the GPU specific use case.

**Comment by damianmoz** [comment]
Does #21813 cover this? Or am being too naive?

**Comment by bradcray** [comment]
> Does https://github.com/chapel-lang/chapel/pull/21813 cover this? Or am being too naive?

It doesn't today (it's limited to the real/uint case only), but it's one of the future potential directions we could take the interface in.

Andy, I'd propose writing this as:

```chapel
var arr64 : [0..<2] uint(64) = [0x1122334455667788, 0xAABBCCDDEEFF0011];
on here.gpus[0] {
  var arr32 : [0..<4] uint(32);
  arr32 = arry4.transmute([0..<4] uint(32));
}
```

or maybe even better:

```chapel
  arr32 = arry4.transmute(arr32.type);
```

or:

```chapel
  var arr32 = arry4.transmute([0..<4] uint(32));
```


My 1.30 release notes slides called this out as a "next step" before your filing of this.

**Comment by mppf** [comment]
I feel that writing this kind of thing with a `memcpy` call would be just fine.

**Comment by stonea** [comment]
> I feel that writing this kind of thing with a `memcpy` call would be just fine.

It's true for the case where we're doing a copy from one CPU-allocated array to another this will work:

```chpl
c_memcpy(c_ptrTo(arr32), c_ptrTo(arr64), arr64.size * numBytes(arr64.eltType))
```

Though I feel it's a little clunky, and I don't think it would work the way I'd want it to in the case where one of the arrays is on the CPU and the other is on the GPU (just tried it out, yes it causes a segfault).

It would also be nice to support something this where no actual data gets copied:

```chpl
const ref B : [0..<4] uint(32) = arr64.transmute([0..<4] uint(32))
```

**Comment by mppf** [comment]
>  I don't think it would work the way I'd want it to in the case where one of the arrays is on the CPU and the other is on the GPU (just tried it out, yes it causes a segfault).

Yes, but that's where a low-level Chapel-tastic `memcpy` / `PUT` / `GET` library routine would come in. Basically, I think that we should have a low level function available that can do this copy regardless of the starting locale of the source and destination.

**Comment by bradcray** [comment]
What Michael's describing here sounds like an extension to https://chapel-lang.org/docs/modules/standard/Communication.html to make it GPU-aware to me.

That said, I also like the idea of having higher-level ways of expressing this through transmute() on arrays (where, ideally, that transmute's implementation would be based on the Communication.get/put calls).

> It would also be nice to support something this where no actual data gets copied:

I also agree with the desire for a no-copy / aliasing array case, but I wouldn't express that with transmute, which—like most (all?) of our casts—creates a new value rather than aliasing an existing one.  Instead, I'd bundle that with the pre-existing and more general desire to create an array using  a pre-existing / borrowed pointer or buffer (which itself might come from applying `c_ptrTo()` to another array).
